go-controller/cmd/ovnkube-trace/ovnkube-trace.go

```diff
@@ -0,0 +1,965 @@
+			Stdin:     useStdin,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, parameterCodec)
+
+	exec, err := remotecommand.NewSPDYExecutor(restconfig, "POST", req.URL())
+	if err != nil {
+		return "", "", err
+	}
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	var stdin io.Reader
+
+	if useStdin {
+		stdin = strings.NewReader(in)
+	} else {
+		stdin = nil
+	}
+
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  stdin,
+		Stdout: &stdout,
+		Stderr: &stderr,
+	})
+	if err != nil {
+		return stdout.String(), stderr.String(), err
+	}
+
+	return stdout.String(), stderr.String(), err
+}
+
+func getPodMAC(client *corev1client.CoreV1Client, pod *kapi.Pod) (podMAC string) {
+
+	if pod.Spec.HostNetwork {
+		node, err := client.Nodes().Get(context.TODO(), pod.Spec.NodeName, metav1.GetOptions{})
+		if err != nil {
+			panic(err)
+		}
+
+		nodeMAC, err := util.ParseNodeManagementPortMACAddress(node)
+		if err != nil {
+			panic(err)
+		}
+		if nodeMAC != nil {
+			podMAC = nodeMAC.String()
+		}
+	} else {
+		podAnnotation, err := util.UnmarshalPodAnnotation(pod.ObjectMeta.Annotations)
+		if err != nil {
+			panic(err)
+		}
+		if podAnnotation != nil {
+			podMAC = podAnnotation.MAC.String()
+		}
+	}
+
+	return podMAC
+}
+
+func getSvcInfo(coreclient *corev1client.CoreV1Client, restconfig *rest.Config, svcName string, ovnNamespace string, namespace string, cmd string) (svcInfo *SvcInfo, err error) {
+
+	// Get service with the name supplied by svcName
+	svc, err := coreclient.Services(namespace).Get(context.TODO(), svcName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(1).Infof("Service %s in namespace %s not found\n", svcName, namespace)
+		return nil, err
+	}
+	klog.V(5).Infof("==>Got service %s in namespace %s\n", svcName, namespace)
+
+	clusterIP := svc.Spec.ClusterIP
+	klog.V(5).Infof("==>Got service %s ClusterIP is %s\n", svcName, clusterIP)
+
+	svcInfo = &SvcInfo{
+		IP: svc.Spec.ClusterIP,
+	}
+
+	fieldSelector := "metadata.name=" + svcName
+	eps, err := coreclient.Endpoints(namespace).List(context.TODO(), metav1.ListOptions{FieldSelector: fieldSelector})
+	if err != nil {
+		klog.V(1).Infof("Endpoints for service %s in namespace %s not found\n", svcName, namespace)
+		return nil, err
+	}
+	klog.V(5).Infof("==>Got EndpointList %v for service %s in namespace %s\n", eps, svcName, namespace)
+
+	// Get one of the podnames that make up the service
+	for _, ep := range eps.Items {
+		klog.V(5).Infof("==>Got endpoint %v for service %s in namespace %s\n", ep, svcName, namespace)
+		for _, subset := range ep.Subsets {
+			klog.V(5).Infof("==>Got subset %v for service %s in namespace %s\n", subset, svcName, namespace)
+			for _, epAddress := range subset.Addresses {
+				klog.V(5).Infof("==>Got Address %v for service %s in namespace %s\n", epAddress, svcName, namespace)
+				svcInfo.PodName = epAddress.TargetRef.Name
+				svcInfo.PodNamespace = epAddress.TargetRef.Namespace
+				break
```

Comment:
This will only break out of inner loop. I.E we will still iterate through all of eps.Items, even if we set `svcInfo.PodName` and `svcInfo.PodNamespace`, Is this what you intended to do? If we just need "One of the podnames" shouldn't we just iterated though eps.Items until we get a valid pod? A way around this would be to encapsulate this into a function and just return once the vars are set